KAFKA-19571: Race condition between log segment flush and file deletion causing log dir to go offline

itoumlilt opened the pull request:
Following JIRA Ticket: https://issues.apache.org/jira/browse/KAFKA-19571

A race condition can occur during replica rebalancing where a log
segment's file is deleted after an asynchronous flush has been scheduled
but before it executes.

This would previously cause an unhandled `ClosedChannelException`,
leading the` ReplicaManager` to mark the entire log directory as
offline.

The fix involves catching the `ClosedChannelException` within the
`LogSegment.flush()` method and suppressing it only if the underlying
log file no longer exists, which is the specific symptom of this race
condition. Legitimate I/O errors on existing files will still be thrown.

Unit test has been added to `LogSegmentTest` to verify both the fix and
the case where the exception should still be thrown.

itoumlilt committed:
FIX: Prevent race condition causing log directory to go offline

A race condition can occur during replica rebalancing where a log segment's
file is closed and deleted/renamed while an asynchronous flush or read
operation is still pending.

This would previously cause an unhandled ClosedChannelException, leading the
ReplicaManager to mark the entire log directory as offline.

The fix involves removing the explicit close() of the source log in
replaceCurrentWithFutureLog(). By leaving the channel open, concurrent
operations can complete successfully on the renamed files (which are moved
to the .delete directory). The log is already scheduled for asynchronous
deletion (via addLogToBeDeleted), which ensures that the log and its
resources will be properly closed and deleted by the background deletion
thread after the configured file delete delay.

A new unit test `testReplaceCurrentWithFutureLogDoesNotCloseSourceLog`
in `LogManagerTest` has been added to verify that the source log is not
closed during the swap operation.

The fix involves removing the explicit close() of the source log in
replaceCurrentWithFutureLog(). By leaving the channel open, concurrent
operations can complete successfully on the renamed files. The log is
already scheduled for asynchronous deletion (via addLogToBeDeleted), which
ensures that the log and its resources will be properly closed and deleted
by the background deletion thread after the configured file delete delay.

itoumlilt commented:
#11438 was fixed to swallow the first ``NoSuchFileException`` WARN in the above stacktrace, but not the underlying exception.
#14280  is similar but different, it swallows `NoSuchFileException` for race condition on log directory move/delete, but not on the segment file level.

junrao commented on storage/src/main/java/org/apache/kafka/storage/internals/log/LogSegment.java:
```diff
@@ -634,6 +635,17 @@ public Void call() throws IOException {
                 }
             });
+        } catch (ClosedChannelException e) {
+            if (!log.file().exists()) {
```
In `replaceCurrentWithFutureLog()`, we rename the log dir to deleted, close the channel and schedule it to be deleted asynchronously. If we get here, it's possible that the renamed file still exists.

Also, while this fixes the issue with flush, the issue with closed channel could be exposed through read too and therefore cause the same issue of forcing the log directory to be offline.

An alternative is to avoid closing `srcLog` in `replaceCurrentWithFutureLog()`. The segments will be closed when the log is deleted after a delay. By that time, the expectation is that there won't be any pending flushes, reads, etc on the log. This approach is also consistent with the approach in `LogManager.asyncDelete()`.

itoumlilt replied:
@junrao  Thanks for the detailed review! You were right: handling this at the `flush()` level was incomplete as it didn't cover read operations.
I've updated the PR to implement your suggestion:
1. Reverted the changes to `LogSegment.java`.
2. Modified `LogManager.replaceCurrentWithFutureLog` to not close the source log immediately. This keeps the channel open for any pending operations on the renamed (.delete) directory.
3. Added a new unit test testReplaceCurrentWithFutureLogDoesNotCloseSourceLog in LogManagerTest to verify that the source log remains open during the swap.

Please let me know if this looks good!

junrao reviewed (commented):
@itoumlilt : Thanks for identifying the issue and submitting a PR. Left a comment.

itoumlilt commented:
Proposed a new fix implementation, amend & push forced to previous commit.
Context : https://github.com/apache/kafka/pull/20289#discussion_r2557245993
This PR is again open to review!

mimaison commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -1249,7 +1247,6 @@ class LogManager(logDirs: Seq[File],
         srcLog.renameDir(UnifiedLog.logDeleteDirName(topicPartition), true)
         // Now that replica in source log directory has been successfully renamed for deletion.
```
If we decide to not close the log here, can we update the comment too? including explaining why

itoumlilt replied:
Done in 19057e4fc8 , updated the comment to explain the reasoning.

mimaison commented on core/src/test/scala/unit/kafka/log/LogManagerTest.scala:
```diff
@@ -43,7 +43,7 @@ import java.util.{Collections, Optional, OptionalLong, Properties}
 import org.apache.kafka.server.storage.log.FetchIsolation
 import org.apache.kafka.server.util.{FileLock, KafkaScheduler, MockTime, Scheduler}
-import org.apache.kafka.storage.internals.log.{CleanerConfig, FetchDataInfo, LogConfig, LogDirFailureChannel, LogMetricNames, LogManager => JLogManager, LogOffsetsListener, LogStartOffsetIncrementReason, ProducerStateManagerConfig, RemoteIndexCache, UnifiedLog}
+import org.apache.kafka.storage.internals.log.{CleanerConfig, FetchDataInfo, LogConfig, LogDirFailureChannel, LogFileUtils, LogMetricNames, LogOffsetsListener, LogStartOffsetIncrementReason, ProducerStateManagerConfig, RemoteIndexCache, UnifiedLog, LogManager => JLogManager}
```
Can we undo the reordering?

itoumlilt replied:
Updated the import reordering in 19057e4fc8

mimaison reviewed (commented):
Thanks for the PR, I left a couple of suggestions

mimaison commented:
I've seen a similar failure.
It's not exactly the same stacktrace as in [KAFKA-19571](https://issues.apache.org/jira/browse/KAFKA-19571), but it seems to be the same underlying issue:
```
ERROR Error while flushing log for <TOPIC> in dir <LOG_DIR> with offset <VALUE> (exclusive) and recovery point <VALUE> (org.apache.kafka.storage.internals.log.LogDirFailureChannel)
java.nio.file.NoSuchFileException: <LOG_DIR>/<TOPIC>-future
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116)
	at java.base/sun.nio.fs.UnixFileSystemProvider.newFileChannel(UnixFileSystemProvider.java:182)
	at java.base/java.nio.channels.FileChannel.open(FileChannel.java:292)
	at java.base/java.nio.channels.FileChannel.open(FileChannel.java:345)
	at org.apache.kafka.common.utils.Utils.flushDir(Utils.java:973)
	at kafka.log.LocalLog.flush(LocalLog.scala:177)
	at kafka.log.UnifiedLog.$anonfun$flush$2(UnifiedLog.scala:1537)
	at kafka.log.UnifiedLog.flush(UnifiedLog.scala:1724)
	at kafka.log.UnifiedLog.flushUptoOffsetExclusive(UnifiedLog.scala:1518)
	at kafka.log.UnifiedLog.$anonfun$roll$1(UnifiedLog.scala:1499)
	at org.apache.kafka.server.util.KafkaScheduler.lambda$schedule$1(KafkaScheduler.java:150)
```

itoumlilt committed:
Address PR feedback: update comment and fix import ordering

itoumlilt commented:
Thanks @mimaison for your review and for sharing this stacktrace! We've encountered the same issue in production and this patch has resolved it for us.

showuon commented on core/src/test/scala/unit/kafka/log/LogManagerTest.scala:
```diff
@@ -1112,6 +1112,41 @@ class LogManagerTest {
+    verify(spyCurrentLog, never()).close()
+
+    // Verify the source log was renamed to .delete
+    assertTrue(spyCurrentLog.dir.getName.endsWith(LogFileUtils.DELETE_DIR_SUFFIX))
```
Could we also verify that in this situation (i.e. after `replaceCurrentWithFutureLog` is invoked without channel closed), the `logSegment.flush()` can be invoked without error?

itoumlilt replied:
Good point, thanks! Added in a1f4547f0d .

showuon reviewed (commented):
Overall LGTM, left a comment to improve the test to make sure when the race condition happened again, the `logSegment.flush()` won't have exception thrown.

itoumlilt committed:
Add flush verification to test race condition fix

Verify that flush() can be called without ClosedChannelException after replaceCurrentWithFutureLog, confirming the race condition is resolved.

showuon commented on core/src/test/scala/unit/kafka/log/LogManagerTest.scala:
```diff
@@ -1112,6 +1112,45 @@ class LogManagerTest {
+    verify(spyCurrentLog, never()).close()
+
+    // Verify the source log was renamed to .delete
+    assertTrue(spyCurrentLog.dir.getName.endsWith(LogFileUtils.DELETE_DIR_SUFFIX))
+
+    // Verify that flush() can be called without error (no ClosedChannelException)
+    val flushLog: Executable = () => spyCurrentLog.flush(false)
+    assertDoesNotThrow(flushLog)
```
To trigger the flush, we have to set `flushOffset > localLog.recoveryPoint()` ([here](https://github.com/apache/kafka/blob/trunk/storage/src/main/java/org/apache/kafka/storage/internals/log/UnifiedLog.java#L2232-L2235)). Because there's no any data, the flushOffset is 0 and recoveryPoint is 0, too. I just tested it, and this flush will not throw exception even if we close the srcLog as before. We have to make the `flushOffset > 0` to trigger the exception, something like this:

```
   // Verify that flush() can be called without error (no ClosedChannelException)
    when(spyCurrentLog.logEndOffset()).thenReturn(100L)
    val flushLog: Executable = () => spyCurrentLog.flush(false)
    assertDoesNotThrow(flushLog)
```

itoumlilt replied:
Indeed, thanks! Updated in 10e33f424d

itoumlilt committed:
Mock logEndOffset to trigger actual flush in test

The flush only occurs when flushOffset > localLog.recoveryPoint()

showuon reviewed (approved):
LGTM!

mimaison reviewed (approved):
Thanks for the PR, LGTM

I was wondering if there was a way to verify that `UnifiedLog.flush()` was actually calling the underlying `flush()` methods on `LocalLog`/`LogSegment` after the `assertDoesNotThrow()` assertion. But I couldn't find a nice way to do so.

mimaison commented:
@junrao Do you want to take another look?

mimaison commented:
Ok I'll go ahead and merge

junrao commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -1247,9 +1247,13 @@ class LogManager(logDirs: Seq[File],
+        // operations (e.g., log flusher, fetch requests) might encounter ClosedChannelException.
+        // The log will be deleted asynchronously by the background delete-logs thread.
+        // File handles are intentionally left open; Unix semantics allow the renamed files
+        // to remain accessible until all handles are closed.
```
>         //The log will be deleted asynchronously by the background delete-logs thread.
>         // File handles are intentionally left open; Unix semantics allow the renamed files
>         // to remain accessible until all handles are closed.

How about the following?

File handles are intentionally left open; Unix semantics allow the renamed files
to remain accessible until all handles are closed. 
The log will be deleted asynchronously by the background delete-logs thread.
File handles are closed and files are deleted after a configured delay log.segment.delete.delay.ms.
At that time, the expectation is that no other concurrent operations need to access
the deleted file handles any more.

junrao reviewed (commented):
@itoumlilt : Thanks for the updated PR. Sorry for the late review. LGTM. Just a minor comment.

junrao commented:
@itoumlilt Could you update the Proposed Fix section in the jira since it's outdated?

@mimaison : Since this is cherry-picked to 4.1, should we cherry-pick this to 4.2 if there is another RC?

mimaison commented:
Yes, we should backport to 4.2. I asked Christo on the dev list to see if we can sneak it in for 4.2.0. Otherwise I'll backport to 4.2 once 4.2.0 is out. I've done the backport to 4.1 yesterday, I'll do it for 4.0 next week (or feel free to do it if you have time)

mimaison commented:
Cherry picked to 4.0, 4.1 and 4.2. The fix will be in 4.0.2, 4.1.2 and 4.2.0
